ambuda/views/auth.py

```diff
@@ -79,51 +91,65 @@ def _is_valid_reset_token(row: db.PasswordResetToken, raw_token: str, now=None):
     return True
 
 
+def get_field_validators(field_name: str, min_len: int, max_len: int):
+    return [
+        val.DataRequired(),
+        val.Length(min=min_len, max=max_len, message=f"{field_name.capitalize()} must be between {min_len} and {max_len} characters long"),
+    ]
+
 class SignupForm(FlaskForm):
-    username = StringField(
-        _l("Username"), [val.Length(min=6, max=25), val.DataRequired()]
+    username = StringField(_l("Username"), [*get_field_validators("username", MIN_USERNAME_LEN, MAX_USERNAME_LEN),
+                                            val.Regexp('^[^\s]*$', message="Username must not contain spaces"), 
+                                            ]
+    )
+    password = PasswordField(_l("Password"), [*get_field_validators("password", MIN_PASSWORD_LEN, MAX_PASSWORD_LEN)])
+    email = StringField(_l("Email address"), [*get_field_validators("email address", MIN_EMAIL_ADDRESS_LEN, MAX_EMAIL_ADDRESS_LEN),
+                                              val.Email(),
+                                             ]
     )
-    password = PasswordField(_l("Password"), [val.Length(min=8), val.DataRequired()])
-    email = StringField(_l("Email address"), [val.DataRequired(), val.Email()])
     recaptcha = RecaptchaField()
 
     def validate_username(self, username):
-        user = q.user(username.data)
+        # username is case insensitive
+        user = q.user(username.data.lower())
         if user:
             raise val.ValidationError("Please use a different username.")
 
     def validate_email(self, email):
         session = q.get_session()
-        user = session.query(db.User).filter_by(email=email.data).first()
+        # email is case insensitive
+        user = session.query(db.User).filter_by(email=email.data.lower()).first()
         if user:
             raise val.ValidationError("Please use a different email address.")
 
 
 class SignInForm(FlaskForm):
-    username = StringField(
-        _l("Username"), [val.Length(min=6, max=25), val.DataRequired()]
+    username = StringField(_l("Username"), [*get_field_validators("username", MIN_USERNAME_LEN, MAX_USERNAME_LEN), 
+                                            val.Regexp('^[^\s]*$', message="Username must not contain spaces"),
+                                            ]
     )
-    password = PasswordField(_l("Password"), [val.Length(min=8), val.DataRequired()])
+    password = PasswordField(_l("Password"), [*get_field_validators("password", MIN_PASSWORD_LEN, MAX_PASSWORD_LEN)])
```

Comment:
since most of these calls are the same, I prefer two separate `get_password_validators` and `get_email_validators` which can be free to call `get_field_validators` underneath.